pkg/controllers/dashboardapi/settings/settings_test.go

```diff
@@ -0,0 +1,263 @@
+	}).AnyTimes()
+
+	alreadyExistsCreateClient.EXPECT().Update(gomock.Any()).DoAndReturn(func(s *v3.Setting) (*v3.Setting, error) {
+		settingsETCD[s.Name] = s
+		return s, nil
+	}).AnyTimes()
+
+	settingsETCD = make(map[string]*v3.Setting)
+
+	err = testSettingsProvider.SetAll(settingMap)
+	assert.Nilf(t, err, "SetAll should not return an error if setting client's Create returns an AlreadyExists error."+
+		" This is because it is assumed that if AlreadyExists is returned, than a different node in the setup created it.")
+}
+
+func populateTestCases() []*testCase {
+	testCases := []*testCase{
+		{
+			description: "test an existing setting with val and empty source being reconfigured with env var uses the value from env var",
+			envVar:      stringPtr("notempty"),
+			newDefVal:   "abc",
+			existingSetting: &v3.Setting{
+				Value:   "somethingelse",
+				Default: "abc",
+			},
+		},
+		{
+			description: "test creating a setting that doesn't exist yet creates the setting in kubernetes",
+			newDefVal:   "abc",
+		},
+		{
+			description: "test changing default of existing setting with a value properly updates the default but doesn't change value",
+			newDefVal:   "newDef",
+			existingSetting: &v3.Setting{
+				Value:   "somethingelse",
+				Default: "oldDef",
+			},
+		},
+		{
+			description: "test changing default of existing setting without a value properly update the default and nothing else",
+			newDefVal:   "newDef",
+			existingSetting: &v3.Setting{
+				Default: "oldDef",
+			},
+		},
+		{
+			description: "test an existing setting with val and \"env\" source being reconfigured with env var updates value to the new env var value",
+			newDefVal:   "abc",
+			envVar:      stringPtr("notempty"),
+			existingSetting: &v3.Setting{
+				Value:   "somethingelse",
+				Default: "abc",
+				Source:  "env",
+			},
+		},
+		{
+			description: "test a setting that doesn't exist with val and \"env\" source being configured with env var creates setting with" +
+				" env var value and \"env\" source",
+			newDefVal: "abc",
+			envVar:    stringPtr("notempty"),
+		},
+		{
+			description: "test that setting an empty string value using an environment variable works when the env var was not used prior",
+			newDefVal:   "abc",
+			envVar:      stringPtr(""),
+			existingSetting: &v3.Setting{
+				Value:   "somethingelse",
+				Default: "abc",
+			},
+		},
+		{
+			description: "test that setting an empty string value using an environment variable works when the env var was used prior.",
+			newDefVal:   "abc",
+			envVar:      stringPtr(""),
+			existingSetting: &v3.Setting{
+				Value:   "somethingelse",
+				Default: "abc",
+				Source:  "env",
+			},
+		},
+	}
+
+	for index, test := range testCases {
+		settingName := fmt.Sprintf("test%d", index)
+		newSetting := settings.NewSetting(settingName, test.newDefVal)
+		test.newSetting = newSetting
+		if test.existingSetting == nil {
```

Comment:
Nit: the if statement with the continue statement could be moved to the top of the loop's body.